spec/requests/api/connect/v3/systems/products_controller_spec.rb

```diff
@@ -207,90 +207,82 @@
-      let(:payload) do
-        {
-          identifier: product.identifier,
-          version: product.version,
-          arch: product.arch
-        }
-      end
+    context 'new product' do
+      subject { response }
 
-      let(:error_response) do
-        {
-          type: 'error',
-          error: "No activation with product '#{product.friendly_name}' was found.",
-          localized_error: "No activation with product '#{product.friendly_name}' was found."
-        }
-      end
+      specify { expect(system.activations.count).to eq(1) }
 
+      its(:code) { is_expected.to eq('201') }
+      its(:body) { is_expected.to eq(serialized_json) }
 
-      its(:code) { is_expected.to eq('422') }
-      its(:body) { is_expected.to eq(error_response.to_json) }
+      it 'activates new product' do
+        expect(system.activations.first.reload.service_id).to equal(new_product.service.id)
+      end
     end
 
-    context 'with activated product' do
-      let(:request) { put url, headers: headers, params: payload }
+    context 'with activated previous product' do
+      subject { response }
 
       let!(:old_product) { FactoryGirl.create(:product, :with_mirrored_repositories, :activated, system: system) }
       let(:new_product) { FactoryGirl.create(:product, :with_mirrored_repositories, predecessors: [old_product]) }
-
-      let(:payload) do
-        {
-          identifier: new_product.identifier,
-          version: new_product.version,
-          arch: new_product.arch
-        }
-      end
-
       let(:serialized_json) do
         V3::ServiceSerializer.new(
           new_product.service,
+          obsoleted_service_name: old_product.service.name,
           base_url: URI::HTTP.build({ scheme: response.request.scheme, host: response.request.host }).to_s
         ).to_json
       end
 
-      describe 'response' do
-        before { request }
-        subject { response }
+      specify { expect(system.activations.count).to eq(1) }
 
-        its(:code) { is_expected.to eq('201') }
-        its(:body) { is_expected.to eq(serialized_json) }
-      end
+      its(:code) { is_expected.to eq('201') }
+      its(:body) { is_expected.to eq(serialized_json) }
 
-      describe 'response with "-" in product version' do
-        before { request }
-        subject { response }
+      it 'deactivates old product and activates new product' do
+        expect(system.activations.first.reload.service_id).to equal(new_product.service.id)
+      end
+    end
 
-        let(:new_product) { FactoryGirl.create(:product, :with_mirrored_repositories, version: '24.0', predecessors: [old_product]) }
-        let(:payload) do
-          {
-            identifier: new_product.identifier,
-            version: '24.0-0',
-            arch: new_product.arch
-          }
-        end
+    context 'with "-" in product version' do
+      before { request }
+      subject { response }
```

Comment:
`subject` in all of those examples is the same and can be moved out to describe block.